README.rst
===================
 django-cairo-text
===================

This is a re-usable Django app for using custom fonts on a web page
for headings, navigation and the like.  It uses the Cairo graphics
library to create images dynamically out of text snippets.  Currently
output in the PNG format is possible since that's all Cairo supports.
You can create ``<img>`` tags which either link to cached image files
or embed encoded image data in the ``src=`` attribute.


Downloads
=========

Use the `django-cairo-text GitHub repository`_ to download the app.

.. _django-cairo-text GitHub repository: http://github.com/akaihola/django-cairo-text/


Dependencies
============

Cairotext uses the PyCairo library so you need to have that and the
Cairo libraries installed along with any fonts you'd like to use.


Quickstart
==========

The template tag ``get_text_image`` generates the image and saves its
URL and pixel size in the context::

  {% load cairotext %}
  {% get_text_image myvar font "Arial" color "#ab3fd0" size 14 as img %}
  Image retrieved with a separate request:
  <img src="{{img.url}}" alt="{{myvar}}"
       width="{{img.width}}" height="{{img.height}}" />
  Image data embedded in HTML:
  <img src="{{img.embed}}" alt="{{myvar}}"
       width="{{img.width}}" height="{{img.height}}" />

You can use the following keywords to customize the appearance of the
text:

* ``font`` (name of typeface)
* ``size`` (font size in pixels)
* ``height`` (image height in pixels)
* ``baseline`` (text baseline in pixels from top of image)
* ``background`` (background color, HTML format)
* ``color`` (text color, HTML format)

Presets can be provided in the Django settings file, and they can be
used and overridden in the template tag.

Behind the scenes, django-cairo-text checks if the image already has
been rendered. If not, a new image is generated and saved to
disk.




Author
======

The original author is `Antti Kaihola`_ of Ambitone Oy.  Ambitone Oy
is a small Finnish new media and web development house which focuses
on Django in its web projects.

.. _Antti Kaihola: http://djangopeople.net/akaihola


Installation
============

Put the cairotext app in your Python path.  Then add 'cairotext' to
INSTALLED_APPS in settings.py.  If you want to customize the defaults,
you can include ``CAIROTEXT_DIR`` (default is ``cairotext_cache``
inside your ``MEDIA_ROOT``) and ``CAIROTEXT_PRESETS`` as well::

  INSTALLED_APPS = (
    # [...]
    'cairotext',)
  CAIROTEXT_DIR = 'cairotext_cache'
  CAIROTEXT_PRESETS = dict(
    fancy=dict(
        font='Arial',
        size=40,
        background='#fc8',
        color='#630',
	height=60,
        baseline=50))


Usage
=====

To show text with a custom font, first load the tag library with ``{%
load cairotext %}`` and then render the image using the
``get_text_image`` template tag::

  {% get_text_image page.title font "jiffy" as img %}

Similarly for string literals::

  {% get_text_image "Hello there!" font "jiffy" as img %}

You can specify font size and color, too::

  {% get_text_image "Hello there!" font "jiffy" size 18 as img %}

Colors are specified with a standard CSS specification. So you can get
green text either::

  {% get_text_image "Hello there!" font "jiffy" color "#00ff00" as img %}

or::

  {% get_text_image "Hello there!" font "jiffy" color "green" as img %}

You can also specify the background color::
  
  {% get_text_image "Hello there!" font "jiffy" color "#0f0" background "black" as img %}

The template tag returns an image object with attributes like this::

  img.url == '/media/cairotext_cache/ba5bfc5b16f25c44bd242230056be56b.png'
  img.embed == (
    'data:image/png;base64,'
    'iVBORw0KGgoAAAANSUhEUgAAAFUAAAAJCAMAAAB0Z5DkAAAAGFBMVEX+/v7+/v79/f3t7e3Hx8eZ\n'
    'mZlSUlIdHR3QDIZLAAAAi0lEQVQoz7VS0Q6DQAgjetD//2OBAjExbkucPFDo2Sp4In8IXBmPF1y/\n'
    'nD9xBUwM+d0KqMCCz1KjVhLdJO3JOKeTOjpiOUGyq5VgbTLlqhTMNGjJsKUjDr2pdzsNa4Isb1zj\n'
    'LCScF6MjhrXFYNyCrHkXy5sN8DGXtGvr1o+/X9HphI/vTSyvNjj4OQ72GgXnYsLNkAAAAABJRU5E\n'
    'rkJggg==')
  img.width == 85
  img.height == 9

Bold or italic text is not directly supported - I haven't looked into
such features in Cairo yet.  You can of course use a bold or italic
version of the font file.

PNG issues
----------

Note that cairotext is able to spit out PNGs with transparency.  To
show these properly in IE 5.5+ and 6, you need to apply the Javascript
IE PNG hack (if you're using a popular Javascript base library,
there's probably a plugin for this).

Caching
-------

If you change the font file or something similar that the caching
mechanism in cairotext is not clever enough to detect, you can simply
delete the generated files. They will then be regenerated. The caching
mechanism doesn't use the database, it only looks at what's available
in the file system.

Presets
-------

In accordance with the DRY principle, you can specify presets in the
Django settings.py file for font settings you use frequently in your
templates.

The ``CAIROTEXT_PRESETS`` dictionary should contain preset names as
keys and rendering option dictionaries as values.  A useful pattern
for inheriting options is to use a function::

  def navi(**kwargs):
      params = dict(
          font='Frutiger LT Std',
          size=14,
          height=14,
          baseline=14,
          color='#5a5a63',
          background='white')
      params.update(kwargs)
      return params
  
  CAIROTEXT_PRESETS = dict(
      navigation=navi(background='#b5b5b5'),
      navigation_hover=navi(background='#b5b5b5', color='white'),
      section_menu=navi(background='#d3d2d2'),
      section_menu_hover=navi(background='#e4e4e4'),
      body_h1=dict(font='Arial', size=16, height=16, baseline=16, color='#505050'))

Use presets in templates by inserting the preset name after the text
to be rendered.  Options can still be overridden::

  {% get_text_image mytext "navigation" color "green" as img %}

This would render green 14px Frutiger on a gray background.

Image optimization
------------------

You might want to optimize the size of generated PNG files in order to
minimize network traffic.  An external optimizer utility can be called
automatically for each generated image.  Two settings have to be
defined:

* CAIROTEXT_OPTIMIZER defines the command line for the optimizer
  binary

* CAIROTEXT_OPTIMIZED_PATH defines where the optimizer saves the
  optimized image

Both settings can contain Python-style named string formatting
specifiers which refer to the following values:

* ``%(path)s``: full path of the final PNG file

* ``%(directory)s``: path of the directory where the PNG is to be
  saved

* ``%(name)s``: image name without the extension

* ``%(ext)s``: always 'png'

An example::

  CAIROTEXT_OPTIMIZER = 'pngnq -s 4 %(path)s'
  CAIROTEXT_OPTIMIZED_PATH = '%(directory)s/%(name)s-nq8.png'

With these settings, Cairotext would optimize each generated file
using the pngnq_ optimizer.  The color palette of the image would be
reduced to four colors.  On one site, this method reduced average
image sizes to 26% of original while still resulting in acceptable
quality.

If the external optimizer returns a non-zero value to indicate an
error, outputs any text on the standard output or standard error
devices or fails to create the optimized image, Cairotext will report
the incident to the site administrators (as set with
``settings.ADMINS``) by e-mail.  In debug mode no e-mail is sent and
an exception is raised instead.

.. _pngnq: http://pngnq.sourceforge.net/


Related apps and blog posts
===========================

Cairotext is originally based on a blog post titled `django and cairo:
rendering pretty titles`_ by Andrew Godwin.  Experience from using and
modifying `django-rendertext`_ (by olau at iola in Denmark, based on
derelm's snippet) and `sorl-thumbnail`_ was also taken advantage of.
Apologies to the django-rendertext team for stealing bits from their
readme file.

.. _django and cairo\: rendering pretty titles: http://www.aeracode.org/2007/12/15/django-and-cairo-rendering-pretty-titles/
.. _django-rendertext: http://code.google.com/p/django-rendertext/
.. _sorl-thumbnail: http://code.google.com/p/sorl-thumbnail/

Other implementations of similar ideas:

* `improved text image view`_ by Jacob Kaplan-Moss
* `simple text image view`_ by derelm, based on work by Andrew Gwozdziewycz and Jacob Kaplan-Moss (above)
* `django-image-replacement`_ by Ludwig Pettersson

.. _improved text image view: http://jacobian.org/writing/improved-text-image-view/
.. _simple text image view:  http://www.djangosnippets.org/snippets/117/
.. _django-image-replacement: http://code.google.com/p/django-image-replacement


CHANGES.rst
==================================
 Change log for django-cairo-text
==================================

-------------------------
Release 1.1.1: 2009-03-20
-------------------------

* a failing optimizer now sends e-mail (production mode) or raises an
  exception (debug mode)
* bugfix: fixed string format in the size mismatch exception message
* minor documentation fix

-----------------------
Release 1.1: 2009-03-16
-----------------------

* added support for external image optimizers
* added support for embedding base64 encoded images in <IMG> tags
* template tag returns a ``TextImage`` instance instead of a dict

-----------------------
Release 1.0: 2009-03-06
-----------------------

* initial release
* renamed ``y`` option to ``baseline``
* removed broken GIF "support"
* added unit tests
* bugfix: cached images not regenerated anymore
* bugfix: copy presets before applying overrides
* documentation: wrote README.rst_
* documentation: wrote CHANGES.rst_

.. _README.rst: README.rst
.. _CHANGES.rst: CHANGES.rst

----------
2009-03-04
----------

* initial checkin to repository
* use PyCairo for rendering
* PNG support
* presets in settings.py
* uploaded to GitHub


cairotext/test_optimizer.py
#!/usr/bin/env python

import sys
from optparse import OptionParser

TEST_IMAGE = (
    '\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00#\x00\x00\x00\r\x08\x06'
    '\x00\x00\x00#\x7f\xd7\x1f\x00\x00\x00\x06bKGD\x00\xff\x00\xff\x00\xff'
    '\xa0\xbd\xa7\x93\x00\x00\x02CIDAT8\x8d\xcd\x931K\xf2Q\x14\xc6\x9f\xfb'
    '\x9a.A4H\x04&\x0e\x0e\x0e\x85DVSC_!P\xa4\xfa\x06-!4\xf8\r\xfa\x00m\xda"'
    '\x14Bm-\xb5\x04\r!\xd4\x14\xce\xd2\xe6\x90R\x1a\x92\xa2\xfd\xfd\xff\xde!'
    '\xbc\xf4\xcf\xbf\xbd\xef\xbb\xc4\xfb\xc0\x19\xee9\xe7\x9e\xe7\xe1\xb9\xf7'
    '\x18\x00c\x8c$\t\xd0O\xc1\x8f\xf3\xd7\x8f\xb1\xff\x05\xfe/1#\xbb\xa4\x0f'
    '\xebF1\x02\xa0\x93\x93\x13mnnjvvV\xc1`P\xf1x\\\x07\x07\x07j\xb7\xdb\xb6o{'
    '{[\xc6\x18%\x93I\xb9\xaek\xf3\xae\xeb*\x99L\xca\x18\xa3\xdd\xdd]\xcb3\x89'
    '\x13\xbf\x00p]\x97\x9d\x9d\x1d\xdf\xba$\x16\x17\x17i\xb7\xdb\x00<??\x13'
    '\x89D\x90D\xb9\\f\x84r\xb9\x8c$b\xb1\x18\xadV\x0b>>\x89\x7f|.~E\xa1P@\x12'
    '\x0b\x0b\x0b\x9c\x9d\x9d\xd1h4\xe8\xf5zT*\x15R\xa9\x14\x92\xc8\xe7\xf3'
    '\xb6\xff\xea\xea\nc\x0c\x89D\x02\xc7qp\x1c\x87D"A \x10\xe0\xe6\xe6\xc63'
    '\xdb\x8f\xf3[1\xeb\xeb\xebH\xe2\xf6\xf6v\xacV\xab\xd5\x90D<\x1e\xf7\xe4'
    '\xf7\xf6\xf6\x90D\xa9T\xa2T*\x8d\t\xfeN\x8c\x81\xc9\xab===\xadn\xb7\xfb'
    '\xed\xa7\x0b\x85B\xea\xf7\xfb\xf6\xfc\xf6\xf6\xa6\xe5\xe5e;kffFwww\n\x06'
    '\x83\x9e{\xff\xbc\xda_\xc5\xf9a0\x18x\xce\x9dNG///z\x7f\x7fW\xb3\xd9T\xbd'
    '^\xd7\xeb\xeb\xeb\x1f\xe7\x8c\x08\x99\x9a\x9aB\x12\x8e\xe3xl[]]E\x12\xf7'
    '\xf7\xf7c6O\xc2\xd6\xd6\x96}\xa6\xc3\xc3C$\x91\xcdf\xc7\xfa\xfc8\x05\x10'
    '\x8dF\x91\xc4\xe5\xe5%\xc3\xe1\xd0\x16\x8f\x8f\x8f\x91\xc4\xfc\xfc<\xc5b'
    '\x91\xc7\xc7G\xba\xdd.\xadV\x8b\x87\x87\x07\x8e\x8e\x8eH\xa5R\xb6\x7f\xb4'
    '9KKK\x0c\x87Cz\xbd\x1e\xb1X\x0cI\x9c\x9f\x9f{\xc4\xf8q\n \x93\xc9L\\\xed'
    '\xfd\xfd\xfd\xc9\xab\xf8\xa9\xf7\xe9\xe9\x89p8\x8c$...,\xe9\xe9\xe9)\x92'
    '\x98\x9b\x9b\xa3\xd9l\xda\xbc\x1f\xa7\x00\xea\xf5:\x99L\x86h4j\xed\xfb'
    '\x8c\xeb\xebk[\x0f\x85B\x84\xc3aVVV\xc8\xe5rT\xabU\x00\xd2\xe94\x92\xd8'
    '\xd8\xd8\xf0\xdcu]\x97\xb5\xb55$\x91N\xa7m\xde\x8f\xf37E\r\xcc\xcc\xd4'
    '\xbf\xd3u\x00\x00\x00\x00IEND\xaeB`\x82')

if __name__ == '__main__':
    p = OptionParser()
    p.add_option('-f', '--fail', action='store_true')
    p.add_option('-n', '--nosave', action='store_true')
    opts, args = p.parse_args()
    if not opts.nosave:
        name, ext = args[0].rsplit('.', 1)
        file('%s-optimized.%s' % (name, ext), 'w').write(TEST_IMAGE)
    sys.exit(bool(opts.fail))


cairotext/templatetags/cairotext.py
from django import template
register = template.Library()

from django.template import Node, TemplateSyntaxError, Variable
from django.core.mail import mail_admins

try:
    from hashlib import md5
except ImportError:
    from md5 import md5
import struct
from urlparse import urljoin
from os import remove, rename
from os.path import join, abspath, exists, split
from pprint import pformat
from subprocess import Popen, PIPE
import cairo

from django.conf import settings

PARAM_KEYWORDS = set(
    ['font', 'size', 'baseline', 'height', 'color', 'background'])

_colors = {
    'aliceblue':'#f0f8ff','antiquewhite':'#faebd7','aqua':'#00ffff',
    'aquamarine':'#7fffd4','azure':'#f0ffff','beige':'#f5f5dc',
    'bisque':'#ffe4c4','black':'#000000','blanchedalmond':'#ffebcd',
    'blue':'#0000ff','blueviolet':'#8a2be2','brown':'#a52a2a',
    'burlywood':'#deb887','cadetblue':'#5f9ea0','chartreuse':'#7fff00',
    'chocolate':'#d2691e','coral':'#ff7f50','cornflowerblue':'#6495ed',
    'cornsilk':'#fff8dc','crimson':'#dc143c','cyan':'#00ffff',
    'darkblue':'#00008b','darkcyan':'#008b8b','darkgoldenrod':'#b8860b',
    'darkgray':'#a9a9a9','darkgreen':'#006400','darkkhaki':'#bdb76b',
    'darkmagenta':'#8b008b','darkolivegreen':'#556b2f','darkorange':'#ff8c00',
    'darkorchid':'#9932cc','darkred':'#8b0000','darksalmon':'#e9967a',
    'darkseagreen':'#8fbc8f','darkslateblue':'#483d8b',
    'darkslategray':'#2f4f4f','darkturquoise':'#00ced1','darkviolet':'#9400d3',
    'deeppink':'#ff1493','deepskyblue':'#00bfff','dimgray':'#696969',
    'dodgerblue':'#1e90ff','firebrick':'#b22222','floralwhite':'#fffaf0',
    'forestgreen':'#228b22','fuchsia':'#ff00ff','gainsboro':'#dcdcdc',
    'ghostwhite':'#f8f8ff','gold':'#ffd700','goldenrod':'#daa520',
    'gray':'#808080','green':'#008000','greenyellow':'#adff2f',
    'honeydew':'#f0fff0','hotpink':'#ff69b4','indianred':'#cd5c5c',
    'indigo':'#4b0082','ivory':'#fffff0','khaki':'#f0e68c',
    'lavender':'#e6e6fa','lavenderblush':'#fff0f5','lawngreen':'#7cfc00',
    'lemonchiffon':'#fffacd','lightblue':'#add8e6','lightcoral':'#f08080',
    'lightcyan':'#e0ffff','lightgoldenrodyellow':'#fafad2',
    'lightgreen':'#90ee90','lightgrey':'#d3d3d3','lightpink':'#ffb6c1',
    'lightsalmon':'#ffa07a','lightseagreen':'#20b2aa','lightskyblue':'#87cefa',
    'lightslategray':'#778899','lightsteelblue':'#b0c4de',
    'lightyellow':'#ffffe0','lime':'#00ff00','limegreen':'#32cd32',
    'linen':'#faf0e6','magenta':'#ff00ff','maroon':'#800000',
    'mediumaquamarine':'#66cdaa','mediumblue':'#0000cd',
    'mediumorchid':'#ba55d3','mediumpurple':'#9370db',
    'mediumseagreen':'#3cb371','mediumslateblue':'#7b68ee',
    'mediumspringgreen':'#00fa9a','mediumturquoise':'#48d1cc',
    'mediumvioletred':'#c71585','midnightblue':'#191970','mintcream':'#f5fffa',
    'mistyrose':'#ffe4e1','moccasin':'#ffe4b5','navajowhite':'#ffdead',
    'navy':'#000080','oldlace':'#fdf5e6','olive':'#808000',
    'olivedrab':'#6b8e23','orange':'#ffa500','orangered':'#ff4500',
    'orchid':'#da70d6','palegoldenrod':'#eee8aa','palegreen':'#98fb98',
    'paleturquoise':'#afeeee','palevioletred':'#db7093','papayawhip':'#ffefd5',
    'peachpuff':'#ffdab9','peru':'#cd853f','pink':'#ffc0cb','plum':'#dda0dd',
    'powderblue':'#b0e0e6','purple':'#800080','red':'#ff0000',
    'rosybrown':'#bc8f8f','royalblue':'#4169e1','saddlebrown':'#8b4513',
    'salmon':'#fa8072','sandybrown':'#f4a460','seagreen':'#2e8b57',
    'seashell':'#fff5ee','sienna':'#a0522d','silver':'#c0c0c0',
    'skyblue':'#87ceeb','slateblue':'#6a5acd','slategray':'#708090',
    'snow':'#fffafa','springgreen':'#00ff7f','steelblue':'#4682b4',
    'tan':'#d2b48c','teal':'#008080','thistle':'#d8bfd8','tomato':'#ff6347',
    'turquoise':'#40e0d0','violet':'#ee82ee','wheat':'#f5deb3',
    'white':'#ffffff','whitesmoke':'#f5f5f5','yellow':'#ffff00',
    'yellowgreen':'#9acd32'}

def convert_color(s):
    if not isinstance(s, basestring):
        return s
    if not s.startswith('#'):
        s = _colors[s]
    l = len(s)
    if l in (4, 5):
        c = [int(x*2, 16)/255.0 for x in s[1:]]
    elif l in (7, 9):
        c = [int(s[i:i+2], 16)/255.0 for i in range(1, l, 2)]
    else:
        raise ValueError('color %r has invalid length' % s)
    if len(c) < 4:
        c.append(1)
    return tuple(c)

def read_png_chunk(pngfile):
    "Read a PNG chunk from the input file, return tag name and data."
    # http://www.w3.org/TR/PNG/#5Chunk-layout
    data_bytes, tag = struct.unpack('!I4s', pngfile.read(8))
    data = pngfile.read(data_bytes)
    return tag, data

def get_png_size(filepath):
    pngfile = file(filepath, 'rb')
    signature = pngfile.read(8)
    if (signature != struct.pack("8B", 137, 80, 78, 71, 13, 10, 26, 10)):
        raise ValueError("Invalid PNG signature")
    while True:
        try:
            tag, data = read_png_chunk(pngfile)
        except ValueError, e:
            raise ValueError('Invalid PNG file: ' + e.args[0])
        if tag == 'IHDR': # http://www.w3.org/TR/PNG/#11IHDR
            return struct.unpack("!2I5B", data)[:2]
    raise ValueError('PNG header not found')

def render_text(text, filepath, params):
    size = params.get('size', 18)
    weight = cairo.FONT_WEIGHT_NORMAL
    style = cairo.FONT_SLANT_NORMAL
    font = params.get('font', 'Sans')

    surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, 1, 1)
    context = cairo.Context(surface)
    context.select_font_face(font, style, weight)
    context.set_font_size(size)
    extents = context.text_extents(text)
    x = -extents[0]
    baseline = params.get('baseline', -extents[1])
    width = extents[2]
    height = params.get('height', extents[3])

    surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, int(width), int(height))
    font_options = surface.get_font_options()
    font_options.set_antialias(cairo.ANTIALIAS_GRAY)
    context = cairo.Context(surface)

    # Paint the background, default is white.
    background = convert_color(params.get('background', (1, 1, 1, 1)))
    context.set_source_rgba(*background)
    context.paint()

    color = convert_color(params.get('color', (0, 0, 0, 1))) # black
    context.set_source_rgba(*color)

    context.select_font_face(font, style, weight)
    context.set_font_size(size)
    # We need to adjust by the text's offsets to center it.
    context.move_to(x, baseline)
    # We stroke and fill to make sure thinner parts are visible.
    context.text_path(text)
    #context.set_line_width(0.05)
    #context.stroke_preserve()
    context.fill()

    try:
        surface.write_to_png(filepath)
        optimizer = Optimizer()
        if optimizer.is_enabled():
            optimizer.optimize(filepath)
    except IOError:
        raise IOError("Can't save image in %r" % abspath(filepath))
    surface.finish()

    return int(width), int(height)

class OptimizerError(Exception): pass

class Optimizer(object):
    def __init__(self):
        self.cmdline_template = getattr(
            settings, 'CAIROTEXT_OPTIMIZER', None)
        self.dest_path_template = getattr(
            settings, 'CAIROTEXT_OPTIMIZED_PATH', None)

    def is_enabled(self):
        return self.cmdline_template and self.dest_path_template

    def optimize(self, filepath):
        self.filepath = filepath
        params = self.get_params_for(filepath)
        self.cmdline = self.cmdline_template % params
        self.dest_path = self.dest_path_template % params
        process = Popen(self.cmdline, shell=True, stdout=PIPE, stderr=PIPE)
        self.stdout, self.stderr = process.stdout.read(), process.stderr.read()
        self.retval = process.wait()
        if self.retval:
            self.error('Cairotext external optimizer failure')
        elif not exists(self.dest_path):
            self.error('Cairotext optimized image missing')
        elif self.stdout or self.stderr:
            self.error('Cairotext optimizer output')
        if exists(self.dest_path):
            # os.rename overwrites existing destination
            rename(self.dest_path, filepath)

    def get_params_for(self, filepath):
        directory, filename = split(filepath)
        name, ext = filename.rsplit('.', 1)
        return {'path': filepath,
                'directory': directory,
                'name': name,
                'ext': ext}

    def error(self, subject):
        message = (
            'Original PNG path: %s\n'
            'Optimized PNG path: %s\n'
            'Optimizer command line: %s\n'
            'Optimizer return value: %d\n%s%s' % (
                self.filepath, self.dest_path, self.cmdline,
                self.retval, self.stdout, self.stderr))
        if settings.DEBUG:
            raise OptimizerError(message)
        else:
            mail_admins(subject, message, fail_silently=True)

class TextImage(object):
    def __init__(self, url, path, size):
        self.url = url
        self.path = path
        self.width, self.height = size
    def _embed(self):
        if not hasattr(self, '_base64'):
            self._base64 = ('data:image/png;base64,%s' %
                            file(self.path).read().encode('base64')[:-1])
        return self._base64
    embed = property(_embed)

class GetTextImageNode(Node):
    def __init__(self, base_params, text, overrides, varname):
        self.base_params = base_params
        self.text = text
        self.overrides = overrides
        self.varname = varname
    def render(self, context):
        params = {}
        if self.base_params is not None:
            params = self.base_params.resolve(context)
        if isinstance(params, basestring):
            try:
                presets = settings.CAIROTEXT_PRESETS
                params = dict(presets[params])
            except (AttributeError, KeyError):
                raise KeyError('Preset "%s" not found in '
                               'settings.CAIROTEXT_PRESETS' % params)
        params.update(dict((key, val.resolve(context))
                           for key, val in self.overrides.items()))
        text = self.text.resolve(context)
        name = md5(text.encode('UTF-8') + pformat(params)).hexdigest()
        render_dir = getattr(settings, 'CAIROTEXT_DIR', 'cairotext_cache')
        filename = '%s.png' % name
        fileurl = urljoin(settings.MEDIA_URL, join(render_dir, filename))
        filepath = join(settings.MEDIA_ROOT, render_dir, filename)
        size = None
        if not exists(filepath):
            size = render_text(text, filepath, params)
        pngsize = get_png_size(filepath)
        assert size is None or size == pngsize, \
            'size mismatch: expected %rx%r, got %rx%r' % (size+pngsize)

        context[self.varname] = TextImage(fileurl, filepath, pngsize)
        return ''

def compile(parser, value):
    if value[0] in '-0123456789':
        return Variable(value)
    return parser.compile_filter(value)

def do_get_text_image(parser, token):
    """
    To use presets from settings.CAIROTEXT_PRESETS['base_params'] and
    override text color, use:

    {% get_text_image "Text" color "#aaa" font "Sans" height 20 as img %}
    <img src="{{img.url}}" width="{{img.width}}" height="{{img.height}}" />
    """
    bits = token.split_contents()
    count = len(bits)
    if count < 4:
        raise TemplateSyntaxError('%r expects at least 3 arguments')
    if bits[-2] != 'as':
        raise TemplateSyntaxError('%r expects "as" as its '
                                  'second last argument')
    text = parser.compile_filter(bits[1])
    base_params = None
    if count % 2:
        base_params = parser.compile_filter(bits[2])
    overrides = dict((keyword, compile(parser, value))
                     for keyword, value in zip(bits[2+count%2:-2:2], bits[3+count%2:-2:2]))
    varname = bits[-1]

    unknown_keywords = set(overrides.keys()).difference(PARAM_KEYWORDS)
    if unknown_keywords:
        raise TemplateSyntaxError('%r got unknown keywords %s' % (
                bits[0],
                ', '.join(unknown_keywords)))
    return GetTextImageNode(base_params, text, overrides, varname)
do_get_text_image = register.tag('get_text_image', do_get_text_image)
